### PR TITLE
[Bug] Update Mistral chat template

### PR DESCRIPTION
**Pull request by riedgar-ms**
It appears that the Mistral chat template has [had an update](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.2/commit/1296dc8fd9b21e6424c9c305c06db9ae60c03ace), so we need to match this

### CODE REVIEW

**Comment by Harsha-Nori** [comment]
I'd also like to think more about if we should force a direct update to the template cache like this, or if we should e.g. have a separate version for the new mistral template update. I'm not sure how frequently model hosts/deployed models will pull new weights/config files down, so this would break them upon a guidance update, while keeping an extra template cache doesn't really cost us anything

File: guidance/chat.py [hunk 1]
```diff
@@ -178,15 +181,15 @@ def get_role_start(self, role_name):
...
 # --------------------------------------------------
 # [05/08/24] https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.2/blob/main/tokenizer_config.json#L42
-mistral_7b_instruct_template = "{{ bos_token }}{% for message in messages %}{% if (message['role'] == 'user') != (loop.index0 % 2 == 0) %}{{ raise_exception('Conversation roles must alternate user/assistant/user/assistant/...') }}{% endif %}{% if message['role'] == 'user' %}{{ '[INST] ' + message['content'] + ' [/INST]' }}{% elif message['role'] == 'assistant' %}{{ message['content'] + eos_token}}{% else %}{{ raise_exception('Only user and assistant roles are supported!') }}{% endif %}{% endfor %}"
+mistral_7b_instruct_template = "{{ bos_token }}{% for message in messages %}{% if (message['role'] == 'user') != (loop.index0 % 2 == 0) %}{{ raise_exception('Conversation roles must alternate user/assistant/user/assistant/...') }}{% endif %}{% if message['role'] == 'user' %}{{ '[INST] ' + message['content'] + ' [/INST]' }}{% elif message['role'] == 'assistant' %}{{ ' ' + message['content'] + eos_token}}{% else %}{{ raise_exception('Only user and assistant roles are supported!') }}{% endif %}{% endfor %}"
```
[the remark this answers is not in the document]
Comment (Harsha-Nori) [reply guidance/chat.py:192]: Ah, didn't see that this was already merged so I just fixed it in https://github.com/guidance-ai/guidance/pull/919

[the remark this answers is not in the document]
Comment (riedgar-ms) [reply guidance/chat.py:192]: Hubris on my part to think that this was a nice, simple fix. I shall give some thought to adding tests to catch such oversights in future

[the remark this answers is not in the document]
Comment (Harsha-Nori) [reply guidance/chat.py:192]: No worries, it was just a single character change! I was able to add it in straight from my phone :)

**Comment by riedgar-ms** [comment]
> I'd also like to think more about if we should force a direct update to the template cache like this, or if we should e.g. have a separate version for the new mistral template update. I'm not sure how frequently model hosts/deployed models will pull new weights/config files down, so this would break them upon a guidance update, while keeping an extra template cache doesn't really cost us anything

I'm not sure how that would work, although learning the details of these chat templates is another fixture on my ToDo list. One immediate question: what if the templates _only_ differed in whitespace (e.g. for this particular one, suppose another update added a second space between the quotes. Since we strip out whitespace when matching keys, we wouldn't be able to detect that.

File: guidance/chat.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Harsha-Nori) [reply guidance/chat.py:192]: Yes absolutely! I'm not even sure all the existing ones are totally correct. When I made these template -> class mappings, I would open up a notebook and check if the guidance classes matched what the Transformers tokenizer's "apply_chat_template" method did on a standard user/assistant "messages" array of dicts. i think formalizing a methodology around that into tests would be incredibly valuable

we could later use that testing infra to start helping us build an auto chat_template loader that we use as a fallback if something isn't in our cache (ie something that tries to convert the chat template into the guidance class format -- if that passes the tests we write with a variety of messages arrays, then we know we're good)

**Comment by Harsha-Nori** [comment]
good q -- I'm really not sure that whitespace stripping the way I did it for cache matching was actually the best decision. jinja ignores some whitespaces but not others, so I figured we'd try to go for maximal coverage at the beginning until figuring out something more clever

**Comment by Harsha-Nori** [comment]
(because the alternative was for the "space misaligned" template to default to the chatml format which would be _**way**_ worse than just matching to something reasonably close...)

**Pull request merged**